### PR TITLE
Move image uploads to queues

### PR DESCRIPTION
**Pull request by mozgzh**
Currently, uploading an avatar means that API has to validate and upload the image to s3 in one request. This can take time and holds up the API, forcing the user to wait until the one request is completed. 

The change is allow the server to handle this whenever it has time through the use of queues. Now the initial upload request returns w/ a 201 and a job id. 

Updates to the front end include waiting for the initial 201 and then polling the /jobs/<id> endpoint to get the status of the upload. Makes use a spinner to indicate loading.

### CODE REVIEW

File: server/package-lock.json [hunk 1]
```diff
@@ -0,0 +1,345 @@
+{
```
Comment (zaccolley) [inline server/package-lock.json:1]: We shouldn't have a package-lock file. Might need to readd this dependancy using yarn.

File: server/package.json [hunk 2]
```diff
@@ -18,6 +18,7 @@
     "db-migrate-plugin-typescript": "2.0.0",
     "event-stream": "3.3.5",
     "express": "4.17.2",
+    "express-json-validator-middleware": "^2.2.1",
```
Comment (zaccolley) [inline server/package.json:21]:
```suggestion
    "express-json-validator-middleware": "2.2.1",
```

File: web/src/components/pages/profile/avatar-setup/avatar-setup.tsx [hunk 3]
```diff
@@ -127,7 +146,20 @@ class AvatarSetup extends React.Component<Props, State> {
       }
     }
 
-    this.setState({ isSaving: false });
+    // this.setState({ isSaving: false });
```
Comment (zaccolley) [inline web/src/components/pages/profile/avatar-setup/avatar-setup.tsx:149]: Commented out code

File: web/src/components/pages/profile/avatar-setup/avatar-setup.tsx [hunk 4]
```diff
@@ -15,6 +15,7 @@ import { Uploads } from '../../../../stores/uploads';
 import { CheckIcon, LinkIcon } from '../../../ui/icons';
 
 import './avatar-setup.css';
+import { Spinner } from '../../../ui/ui';
```
Comment (zaccolley) [inline web/src/components/pages/profile/avatar-setup/avatar-setup.tsx:18]: Style - not important, could we move these dependancies above the CSS import with the others?

File: server/src/server.ts [hunk 5]
```diff
@@ -26,7 +25,7 @@ import authRouter from './auth-router';
 import fetchLegalDocument from './fetch-legal-document';
 import { createTaskQueues, TaskQueues } from './lib/takeout';
 import getCSPHeaderValue from './csp-header-value';
-
+const { ValidationError } = require('express-json-validator-middleware');
```
Comment (zaccolley) [inline server/src/server.ts:28]: Not too importnat, can we use `import` here?

File: server/src/server.ts [hunk 6]
```diff
@@ -155,21 +155,29 @@ export default class Server {
...
-        const isAPIError = error instanceof APIError;
-        if (!isAPIError) {
-          console.error(request.url, error.message, error.stack);
+      app.use(
+        (
+          error: any,
+          request: Request,
+          response: Response,
+          // eslint-disable-next-line @typescript-eslint/no-unused-vars
+          _next: NextFunction // this unused parameter must be included for error handling middleware
```
Comment (zaccolley) [inline server/src/server.ts:164]: Nice comment thanks

File: server/src/lib/queues/processes/imageProcessor.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
+import { Job } from 'bull';
+import Bucket from '../../bucket';
+// import { getMySQLInstance } from '../../model/db/mysql';
```
Comment (zaccolley) [inline server/src/lib/queues/processes/imageProcessor.ts:3]: Commented out code

File: server/src/lib/queues/imageQueue.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
+// import { S3 } from 'aws-sdk';
```
Comment (zaccolley) [inline server/src/lib/queues/imageQueue.ts:1]: Commented out code

File: server/src/lib/queues/imageQueue.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+import * as Queue from 'bull';
+import { getConfig } from '../../config-helper';
+// import Bucket from '../bucket';
+import imageProcessor from './processes/imageProcessor'; // producer
+
+const redisUrlParts = getConfig().REDIS_URL?.split('//');
```
Comment (zaccolley) [inline server/src/lib/queues/imageQueue.ts:7]: Style not important, could we put all of this logic in a 'getRedisOptions' function or something?

File: server/src/lib/api.ts [hunk 10]
```diff
@@ -515,6 +485,25 @@ export default class API {
     response.json({ url });
   };
 
+  getJob = async (
+    { client_id, params: { jobId } }: Request,
+    response: Response,
+    next: NextFunction
+  ) => {
+    // if (!isValidJobId(jobId)) throw new APIError('Invalid job request');
```
Comment (zaccolley) [inline server/src/lib/api.ts:493]: Commented out code

File: server/package.json (hunk 2 above)
Comment (mozgzh) [reply server/package.json:21]: We don't need to lock in dependencies do we?

Comment (zaccolley) [reply server/package.json:21]: Renovate is configured to lock dependancies and automerge anything but major (breaking) changes. So this just saves renovate from opening a PR to pin this.

**Review by zaccolley — APPROVED** [review]

**Pull request merged**